### PR TITLE
Generate test run reports using mako/bokeh.

### PR DESCRIPTION
**Pull request by doctoryes**
Enhances the test run reports with graphs for every request type on a generated HTML page with the Locust data in tables.

### CODE REVIEW

File: util/csm_reports.py [hunk 1]
```diff
@@ -12,40 +14,20 @@
 # TODO: Figure out why...
 import pandas as pd
 import numpy as np
```
Comment (macdiesel) [inline util/csm_reports.py]: nit: would it be better to use the actual library names?

File: util/csm_reports.py [hunk 2]
```diff
@@ -12,40 +14,20 @@
...
-DEFAULT_MONGO_PORT = 27017
-DEFAULT_MONGO_DATABASE = "locust_data"
+from bokeh.plotting import figure as bokeh_figure, output_file as bokeh_output_file, show as bokeh_show
+from bokeh.embed import components as bokeh_components
+from mako.template import Template
 
-RAW_DATA_COLLECTION_FMT = 'requests_{}'
-TEST_RUN_COLLECTION = "test_runs"
+# Work around the fact that this code doesn't live in a proper Python package.
+sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'locust', 'helpers'))
```
Comment (macdiesel) [inline util/csm_reports.py]: should we put this inside a proper python package?

File: util/csm_reports.py [hunk 3]
```diff
@@ -86,43 +71,116 @@ def scatter_plot(successes, failures):
...
-    plt.xlim(min_time, max_time)
+    TOOLS = "resize,crosshair,pan,wheel_zoom,box_zoom,reset,box_select"
+
+    # Find the maximum y-value.
+    y_max = max(max_per_interval)
+    if len(failure_resp_times):
+        y_max = max(y_max, max(failure_resp_times))
+
+    # Create a new plot with the tools above, and set axis labels.
+    p = bokeh_figure(
```
Comment (macdiesel) [inline util/csm_reports.py]: nit: more descriptive variable names would help readability.

**Comment by macdiesel** [comment]
few nits but LGTM :+1: can't wait to see the report!

File: util/csm_reports.py (hunk 2 above)
Comment (doctoryes) [inline util/csm_reports.py]: This formulation is better than code duplication (which is what I'd done before).
The stuff in helpers is teetering on the edge of possibility as a module. But there's not a lot in there yet. In the private repo, there's not much there either. I'm leaning towards leaving it for now.

File: util/csm_reports.py (hunk 3 above)
Comment (doctoryes) [inline util/csm_reports.py]: Agreed! I'll change it.

**Comment by doctoryes** [comment]
Here's the latest state of raw data collection / report generation.
@macdiesel  - Made some additional commits since your thumb.
@alawibaba @cpennington - Please review.
Change and/or merge this as you see fit.

**Comment by cpennington** [comment]
@alawibaba @macdiesel: Can I get a second set of eyes on this, now that I've made a bunch of modifications? Especially on `raw_logs.py`.

We're going to need a way to integrate the CSM reporting with the new log format, but that should be a relatively easy tweak (that'll be my next step)

**Comment by alawibaba** [comment]
:+1: I have a couple of nits I'll inline.

File: util/csm_reports.py [hunk 4]
```diff
@@ -86,56 +78,196 @@ def scatter_plot(successes, failures):
...
+        (successes, failures) = data_source.get_req_data(req_type)
+        if len(successes):
+            all_plots.append(scatter_plot(
+                successes, failures, label=req_type, min_time=min_time, max_time=max_time
+            ))
+
+    # Output an HTML report of the test run.
+    script, divs = bokeh_components(all_plots)
+
+    with open('report.html', 'w') as outfile:
```
Comment (alawibaba) [inline util/csm_reports.py]: Please make this am argument to the method.

**Pull request merged**